solver/app/quote.go

```diff
@@ -1,63 +1,217 @@
+		}
+
+		depositTkn, ok := tokens.find(req.SourceChainID, req.DepositToken)
+		if !ok {
+			return Deposit{}, newRejection(rejectUnsupportedDeposit, errors.New("unsupported deposit token", "addr", req.DepositToken))
+		}
+
+		var expenses []Payment
+		for _, e := range req.Expenses {
+			tkn, ok := tokens.find(req.DestinationChainID, e.Token)
+			if !ok {
+				return Deposit{}, newRejection(rejectUnsupportedExpense, errors.New("unsupported expense token", "addr", e.Token))
+			}
+			expenses = append(expenses, Payment{
+				Token:  tkn,
+				Amount: e.Amount,
+			})
+		}
+
+		quote, err := getQuote([]Token{depositTkn}, expenses)
+		if err != nil {
+			return Deposit{}, err
+		}
+
+		if err := checkLiquidity(ctx, expenses, backend, solverAddr); err != nil {
+			return Deposit{}, err
+		}
+
+		return Deposit{
+			Token:  quote[0].Token.Address,
+			Amount: quote[0].Amount,
+		}, nil
+	}
+}
+
+// newQuoteHandler returns a handler for the /quote endpoint.
+// It is responsible to http request / response handling, and delegates
+// logic to a quoteFunc.
+func newQuoteHandler(quoteFunc quoteFunc) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, rr *http.Request) {
+		ctx := rr.Context()
+
+		w.Header().Set("Content-Type", "application/json")
+
+		// TODO: better request / response logging
+
+		write := func(res QuoteResponse) {
+			// Write response header first, before body.
+			w.WriteHeader(res.StatusCode())
+			if err := json.NewEncoder(w).Encode(res); err != nil {
+				log.Error(ctx, "[BUG] error writing /quote response", err)
+			}
+		}
+
+		writeError := func(statusCode int, err error) {
+			log.DebugErr(ctx, "Error handling /quote request", err)
+
+			write(QuoteResponse{
+				Error: &Error{
+					Code:    statusCode,
+					Status:  http.StatusText(statusCode),
+					Message: err.Error(),
+				},
+			})
+		}
+
+		var req QuoteRequest
+		if err := json.NewDecoder(rr.Body).Decode(&req); err != nil {
+			writeError(http.StatusBadRequest, errors.Wrap(err, "decode request"))
+			return
+		}
+
+		deposit, err := quoteFunc(ctx, req)
+		if r := new(RejectionError); errors.As(err, &r) { // RejectionError
+			write(QuoteResponse{
+				Rejected:          true,
+				RejectReason:      r.Reason.String(),
+				RejectDescription: r.Err.Error(),
+			})
+		} else if err != nil { // Error
+			writeError(http.StatusInternalServerError, err)
+		} else {
+			write(QuoteResponse{Deposit: &deposit}) // Success
+		}
+	})
+}
+
 // getQuote returns payment in `depositTkns` required to pay for `expenses`.
 //
 // For now, this is a simple quote that requires a single expense, paid
 // for by an equal amount of an equivalent deposit token. Token equivalence is
 // determined by symbol (ex arbitrum "ETH" is equivalent to optimism "ETH").
-func getQuote(depositTkns []Token, expenses []Payment) ([]Payment, RejectOrErr) {
+func getQuote(depositTkns []Token, expenses []Payment) ([]Payment, error) {
```

Comment:
refactored `RejectOrError` into standard `error`:
 - Return `RejectionError` which is a rejection
 - Or return `errors.New` which is a error
 
 Handle using `errors.As`